salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,835 @@
+            kind=kind,
+            plural=plural,
+            scope=scope
+        )
+        self._kind = kind
+
+    api_version = property(operator.attrgetter('_api_version'))
+    kind = property(operator.attrgetter('_kind'))
+    scope = property(operator.attrgetter('_scope'))
+    client = property(operator.attrgetter('_client'))
+
+    @property
+    def key(self):
+        # Used for indexing in `KNOWN_CUSTOM_KINDS`
+        return (self.api_version, self.kind)
+
+
+_CUSTOM_KINDS = [
+    CRKindInfo('monitoring.coreos.com/v1', 'Alertmanager',
+               scope='namespaced', plural='alertmanagers'),
+    CRKindInfo('monitoring.coreos.com/v1', 'Prometheus',
+               scope='namespaced', plural='prometheuses'),
+    CRKindInfo('monitoring.coreos.com/v1', 'PrometheusRule',
+               scope='namespaced', plural='prometheusrules'),
+    CRKindInfo('monitoring.coreos.com/v1', 'ServiceMonitor',
+               scope='namespaced', plural='servicemonitors'),
+]
+
+KNOWN_CUSTOM_KINDS = {kind.key: kind for kind in _CUSTOM_KINDS}
+
+
+class _DictWrapper(object):
+    """Wrapper for dynamic attribute access support over a simple dict.
+
+    Implemented by recursively casting all dicts within the original dict as
+    other `_DictWrapper`s, and overloading the `__getattribute__` to fallback
+    on the internal `_fields` for retrieving an attribute.
+
+    Used by `CustomObject` to emulate the other models for `kubernetes.client`.
+
+    >>> example = _DictWrapper({
+    ...     'a': {'b': 'value'},
+    ...     'c': [{'subkey': 'subvalue'}]
+    ... })
+    >>> example.a.b
+    'value'
+    >>> example.c[0].subkey
+    'subvalue'
+    >>> example.unknown
+    Traceback (most recent call last):
+        ...
+    AttributeError: Custom object has no attribute 'unknown'
+    """
+    def __init__(self, fields):
+        self._fields = fields
+
+    @classmethod
+    def from_value(cls, value):
+        if isinstance(value, dict):
+            return cls(value)
+        if isinstance(value, list):
+            return [cls.from_value(val) for val in value]
+        return value
+
+    def __repr__(self):
+        return repr(self._fields)
+
+    def __getattribute__(self, name):
+        try:
+            return super(_DictWrapper, self).__getattribute__(name)
+        except AttributeError:
+            if name in self._fields:
+                return self.from_value(self._fields[name])
+            for key in self._fields:
+                if _convert_attribute_name(key) == name:
+                    return self.from_value(self._fields[key])
+            raise AttributeError(
+                "Custom object has no attribute '{}'".format(name)
+            )
+
+    def __setattr(self, name, value):
+        # First check for class values then retrieve from dict
+        if name in ['_fields']:
+            super(_DictWrapper, self).__setattr__(name, value)
+        else:
+            self._fields[name] = value
```

Comment:
What's that? I'm not sure we want that kind of support, but if we do, the name of the method should be `__setattr__`, and the first check should be `if name not in self._fields` IMHO. Also, doctests should be updated.

In any case, given the way we have a read-only `_body` in `CustomObject`, I really don't think it's going to be useful.

Reply:
> the name of the method should be `__setattr__`

Oh yes for the name of the function. Typo during my copy/paste test :stuck_out_tongue:.

> and the first check should be `if name not in self._fields`

Hmm, no if we want to add a new value the value will not be in `_fields` 

> In any case, given the way we have a read-only `_body` in `CustomObject`, I really don't think it's going to be useful.

It's needed we can't have a read only body for replace we need to set the `resourceVersion` to the one of the old object so either we handle directly dict everywhere and we can basically get rid of `CustomObjects` or we make `CustomObjects` editable.